### ISSUE TITLE
System.InvalidOperationException: Invalid attempt made to decrement the event's count below zero.

### ISSUE THREAD

**Issue by RalfKoban** [nunit/nunit#3273]
Hi,

I'm using NUnit 3.12 and NCrunch 3.28.0.6 on a .NET Framework 4.8 C# solution in VS 2017 (v15.9.12).
NCrunch is running in Auto-churn mode - that is, the tests get executed in an endless loop.

After some time (maybe half a minute or so) it happens that some of the tests fail with following exception:

``` C#
An exception was thrown in the Task Environment: System.InvalidOperationException: Invalid attempt made to decrement the event's count below zero.
   at System.Threading.CountdownEvent.Signal()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.OnChildItemCompleted(Object sender, EventArgs e) in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs:line 374
   at System.EventHandler.Invoke(Object sender, EventArgs e)
   at NUnit.Framework.Internal.Execution.WorkItem.WorkItemComplete() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 378
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItem.cs:line 79
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 521
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs:line 296
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 521
   at NUnit.Framework.Internal.Execution.TestWorker.TestWorkerThreadProc() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\TestWorker.cs:line 142
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
System.InvalidOperationException: Invalid attempt made to decrement the event's count below zero.
   at System.Threading.CountdownEvent.Signal()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.OnChildItemCompleted(Object sender, EventArgs e) in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs:line 374
   at System.EventHandler.Invoke(Object sender, EventArgs e)
   at NUnit.Framework.Internal.Execution.WorkItem.WorkItemComplete() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 378
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItem.cs:line 79
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 521
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs:line 296
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 521
   at NUnit.Framework.Internal.Execution.TestWorker.TestWorkerThreadProc() in D:\a\1\s\src\NUnitFramework\framework\Internal\Execution\TestWorker.cs:line 142
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

Unfortunately, I don't have any example code but it seems like there is a race-condition or something inside CompositeWorkItem.cs, because always other tests fail (and the longer the "endless loop" runs, the more tests fail with the same exception).

BTW: In older versions of NUnit (e.g. v3.9) I did not see those exceptions.

Best regards,
Ralf

**Comment by rprouse** [comment]
Thanks, we'll see if we can reproduce in our unit tests.

**Comment by CharliePoole** [comment]
@rprouse This is very believable. I'm working on a separate issue and I think we are hurt by the fact that we don't actively track test fixtures and suites (CompositeWorkItems) that are waiting for their child tests to run.

I initially wrote the code to use simple CountdownEvents to trigger one-time teardown. My thought was, "well, this probably won't work, but it's the simplest thing so let's try." It turned out to work so I left it in with the notion that we might replace it with more active tracking of suite status when it became necessasry.

My analysis now is that this is a good approach only if __everyone knows about it__, which was obviously not the case. New features got added and I think the original approach was outgrown without our realizing it.

You probably won't want to make a substantive change to the architecture for this issue alone, but I think that fixtures waiting for their children to finish will probably need to go into a special holding queue managed by the ParallelWorkItemDispatcher. It's a rather big change to contemplate, however.

**Comment by mikkelbu** [comment]
Seems like this were closed in version 3.13.3 by #3903. So closing this issue
